src/map.cpp

```diff
@@ -3017,315 +3017,300 @@ ter_id map::get_roof( const tripoint &p, const bool allow_air )
-                if( bash->bash_below && ter_below.has_flag( "SUPPORTS_ROOF" ) ) {
-                    // When bashing the tile below, don't allow bashing the floor
-                    bash_ter_furn( below, str, silent, destroy, false, res_roll );
+                if ( bash->str_max_supported != -1 ) {
+                    smax = bash->str_max_supported;
                 }
+            }
+        }
+        // Linear interpolation from str_min to str_max
+        const int resistance = smin + ( params.roll * ( smax - smin ) );
+        if( params.strength >= resistance ) {
+            success = true;
+        }
+    }
+
+    if( !params.destroy && !success ) {
+        if( sound_fail_vol == -1 ) {
+            sound_volume = 12;
+        } else {
+            sound_volume = sound_fail_vol;
+        }
+
+        sound = _(bash->sound_fail.c_str());
+        params.did_bash = true;
+        if( !params.silent ) {
+            sounds::sound( p, sound_volume, sound, false, "bash", sound );
+        }
+
+        return;
+    }
+
+    // Clear out any partially grown seeds
+    if( has_flag_ter_or_furn( "PLANT", p ) ) {
+        i_clear( p );
+    }
+
+    if( ( smash_furn && has_flag_furn("FUNGUS", p) ) ||
+        ( smash_ter && has_flag_ter("FUNGUS", p) ) ) {
+        create_spores( p );
+    }
 
-                ter_set( p, t_open_air );
+    if( params.destroy ) {
+        sound_volume = smax;
+    } else {
+        if( sound_vol == -1 ) {
+            sound_volume = std::min(int(smin * 1.5), smax);
+        } else {
+            sound_volume = sound_vol;
+        }
+    }
+
+    sound = _(bash->sound.c_str());
+    // Set this now in case the ter_set below changes this
+    const bool collapses = has_flag("COLLAPSES", p) && smash_ter;
+    const bool supports = has_flag("SUPPORTS_ROOF", p) && smash_ter;
+
+    const bool tent = smash_furn && !bash->tent_centers.empty();
+    // Special code to collapse the tent if destroyed
+    if( tent ) {
+        // Get ids of possible centers
+        std::set<furn_id> centers;
+        for( const auto &center : bash->tent_centers ) {
+            const furn_id cur_id = furnfind( center );
+            if( cur_id != f_null ) {
+                centers.insert( cur_id );
             }
+        }
 
-            if( smash_ter && ter( p ) == t_open_air ) {
-                if( !zlevels ) {
-                    // We destroyed something, so we aren't just "plugging" air with dirt here
-                    ter_set( p, t_dirt );
-                } else {
-                    tripoint below( p.x, p.y, p.z - 1 );
-                    const auto roof = get_roof( below, bash_floor && ter_at( below ).movecost != 0 );
-                    ter_set( p, roof );
+        tripoint tentp = tripoint_min;
+        furn_id center_type = f_null;
+        
+        // Find the center of the tent
+        // First check if we're not currently bashing the center
+        if( centers.count( furn( p ) ) > 0 ) {
```

Comment:
Why is this case checked separately? Isn't the center point included in the points returned by `points_in_radius`? I think the scenario where someone bashes the center of the tent is not much more likely then bashing the outsides of it. Checking this separately seems (in my opinion) not useful.